index.js

```diff
@@ -3044,81 +3044,151 @@
     impl: at
   };
 
-  //# head :: Array a -> Maybe a
+  //# head :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the first element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the first element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.head ([1, 2, 3])
   //. Just (1)
   //.
   //. > S.head ([])
   //. Nothing
+  //.
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. S.Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. S.Nothing
   //. ```
+
+  function head(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isJust ? maybe : Just (x);
+      },
+      Nothing,
+      foldable
+    );
+  }
   _.head = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
-    impl: array (Nothing) (B (K) (Just))
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
+    impl: head
+
   };
 
-  //# last :: Array a -> Maybe a
+  //# last :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the last element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the last element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.last ([1, 2, 3])
   //. Just (3)
   //.
   //. > S.last ([])
   //. Nothing
+  //.
+  //. > S.last (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))
+  //. S.Just ('baz')
+  //.
+  //. > S.last (Nil)
+  //. S.Nothing
   //. ```
-  function last(xs) {
-    return xs.length > 0 ? Just (xs[xs.length - 1]) : Nothing;
+
+  function last(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return Just (x);
+      },
+      Nothing,
+      foldable
+     );
   }
+
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Applicative f, Foldable f, Monoid f) => f a -> Maybe (f a)
   //.
-  //. Returns Just all but the first of the given array's elements if the
-  //. array contains at least one element; Nothing otherwise.
+  //. Returns a Just of all but the first of the given structure if
+  //. it contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.tail ([1, 2, 3])
   //. Just ([2, 3])
   //.
   //. > S.tail ([])
   //. Nothing
+  //.
+  //. > S.tail (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))
+  //. Just (Cons ('bar') (Cons ('baz') (Nil)))
+  //
+  //. > S.tail (Nil)
+  //. Nothing
   //. ```
+
+  function tail(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isNothing ?
+               Just (Z.empty (foldable.constructor)) :
+               Z.map (append (x), maybe);
+      },
+      Nothing,
+      foldable
+    );
+  }
+
```

Comment:
The empty lines before and after the implementations can be removed.

Reply:
🆗